crates/win-api-wrappers/src/service.rs

```diff
@@ -0,0 +1,142 @@
+use windows::core::Owned;
+use windows::Win32::Foundation::{ERROR_INSUFFICIENT_BUFFER, GENERIC_READ};
+use windows::Win32::System::Services::{
+    OpenSCManagerW, OpenServiceW, QueryServiceConfigW, QueryServiceStatus, StartServiceW, QUERY_SERVICE_CONFIGW,
+    SC_HANDLE, SC_MANAGER_ALL_ACCESS, SERVICE_ALL_ACCESS, SERVICE_AUTO_START, SERVICE_BOOT_START, SERVICE_DEMAND_START,
+    SERVICE_DISABLED, SERVICE_QUERY_CONFIG, SERVICE_QUERY_STATUS, SERVICE_RUNNING, SERVICE_STATUS,
+    SERVICE_SYSTEM_START,
+};
+
+use crate::utils::WideString;
+use crate::Error;
+
+pub struct ServiceManager {
+    handle: Owned<SC_HANDLE>,
+}
+
+impl ServiceManager {
+    pub fn open_read() -> Result<Self, Error> {
+        Self::open_with_access(GENERIC_READ.0)
+    }
+
+    pub fn open_all_access() -> Result<Self, Error> {
+        Self::open_with_access(SC_MANAGER_ALL_ACCESS)
+    }
+
+    fn open_with_access(access: u32) -> Result<Self, Error> {
+        // SAFETY: No preconditions.
+        let raw_sc_handle = unsafe { OpenSCManagerW(None, None, access)? };
+
+        // SAFETY: No preconditions.
+        let handle = unsafe { Owned::new(raw_sc_handle) };
+
+        Ok(Self { handle })
+    }
+
+    fn open_service_with_access(&self, service_name: &str, access: u32) -> Result<Service, Error> {
+        let service_name_wide = WideString::from(service_name);
+
+        // SAFETY: No preconditions.
+        let raw_service_handle = unsafe { OpenServiceW(*self.handle, service_name_wide.as_pcwstr(), access)? };
+
+        // SAFETY: No preconditions.
+        let handle = unsafe { Owned::new(raw_service_handle) };
+
+        Ok(Service { handle })
+    }
+
+    pub fn open_service_read(&self, service_name: &str) -> Result<Service, Error> {
+        self.open_service_with_access(service_name, SERVICE_QUERY_CONFIG | SERVICE_QUERY_STATUS)
+    }
+
+    pub fn open_service_all_access(&self, service_name: &str) -> Result<Service, Error> {
+        self.open_service_with_access(service_name, SERVICE_ALL_ACCESS)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ServiceStartupMode {
+    Boot,
+    System,
+    Automatic,
+    Manual,
+    Disabled,
+}
+
+pub struct Service {
+    handle: Owned<SC_HANDLE>,
+}
+
+impl Service {
+    pub fn startup_mode(&self) -> Result<ServiceStartupMode, Error> {
+        let mut cbbufsize = 0u32;
+        let mut pcbbytesneeded = 0u32;
+
+        // SAFETY: No preconditions. Query required buffer size.
+        let result = unsafe { QueryServiceConfigW(*self.handle, None, 0, &mut pcbbytesneeded) };
+
+        match result {
+            Err(err) if err.code() == ERROR_INSUFFICIENT_BUFFER.to_hresult() => {
+                // Expected error, continue.
+            }
+            Err(err) => return Err(err.into()),
+            Ok(_) => panic!("QueryServiceConfigW should fail with ERROR_INSUFFICIENT_BUFFER"),
+        }
+
+        // Event though `QueryServiceConfigW` states that `lpserviceconfig` is a `buffer`,
```

Comment:
note: FYI, the terminology "buffer" used in Windows doc is a bit different to what you may be expecting. The most typical buffer we work with in Rust are arrays of bytes such as `[u8]`, `Vec<u8>`. In Microsoft’s Win32 documentation, a “buffer” generally refers to a caller-allocated memory region that an API uses to either input or output data, and it is ultimately coerced into some other type with various alignment requirements. Feel free to reuse this verbatim in your comment if you find this notion useful.